src/util/util.hpp

```diff
@@ -231,14 +229,29 @@ class ExpirationCache {
  public:
   ExpirationCache(uint32_t max_size, uint32_t delete_step) : max_size_(max_size), delete_step_(delete_step) {}
 
+  /**
+   * @brief Inserts key into the cache map. In case provided key is already in cache, only shared lock
+   *        is acquired and function returns false. This means insert does not need to be used together with count()
+   *        to save the performance by not acquiring unique lock
+   *
+   * @param key
+   * @return true if actual insertion took place, otherwise false
+   */
   bool insert(Key const &key) {
-    boost::unique_lock lck(mtx_);
+    {
+      boost::shared_lock lock(mtx_);
+      if (cache_.count(key)) {
+        return false;
+      }
+    }
 
-    if (cache_.find(key) != cache_.end()) {
+    boost::unique_lock lock(mtx_);
+
+    // There must be double check if key is not already in cache due to possible race condition
+    if (!cache_.insert(key).second) {
```

Comment:
you can get rid of this with upgradable lock

Reply:
nope, I did check boost documentation properly and there is a huge difference between upgradable + upgrade_to_unique lock vs sharedlock + unique lock. ONLY 1 SINGLE UPGRADABLE LOCK AT THE SAME TIME CAN BE ACQUIRED, so if this is the method that is called very often, it behaves almost as unique_lock. And because this is called very often, it would block other threads even in case some element is already in cache. In shared_lock + unique_lock if the element is already in cache it will not block other threads because  you can acquire as many shared_locks as you wish at the same time

Reply:
but with this there is edge scenario when you pass first part of scope and then other thread calls count() and return size-1 as you did not do insert meanwhile

Reply:
I dont understand what scenario are you talking about ? but I dont think there is any edge case, insertion is guarded by unique_lock. The first part with shared_lock is purely as performance optimization, it basically does not even need to be there and it would behave in the same way but worse performance